gapis/api/subcmd_idx_trie.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (C) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+// SubCmdIdxTrie is a map-based trie using SubCmdIdx for indexing the data
+// stored inside.
+type SubCmdIdxTrie struct {
+	value     interface{}
+	nextLevel map[uint64]*SubCmdIdxTrie
+}
+
+func (t *SubCmdIdxTrie) index(indices SubCmdIdx, indexCb func(*SubCmdIdxTrie),
+	preOrderCb, postOrderCb func(*SubCmdIdxTrie, SubCmdIdx)) bool {
+	if len(indices) == 0 {
+		indexCb(t)
+		return true
+	}
+	var ret bool
+	if preOrderCb != nil {
+		preOrderCb(t, indices)
+	}
+	if next, ok := t.nextLevel[indices[0]]; ok {
+		ret = next.index(indices[1:], indexCb, preOrderCb, postOrderCb)
+	} else {
+		return false
+	}
+	if postOrderCb != nil {
+		postOrderCb(t, indices)
+	}
+	return ret
+}
+
+// Value returnes the value stored in the trie indexed by the given SubCmdIdx.
+// if no value is found by the given SubCmdIdx, returns nil.
+func (t *SubCmdIdxTrie) Value(indices SubCmdIdx) interface{} {
+	var n *SubCmdIdxTrie
+	t.index(indices, func(t *SubCmdIdxTrie) { n = t }, nil, nil)
+	if n != nil {
+		return n.value
+	}
+	return nil
+}
+
+// SetValue sets a value to the trie with the given SubCmdIdx as index.
+func (t *SubCmdIdxTrie) SetValue(indices SubCmdIdx, v interface{}) {
+	t.index(indices, func(t *SubCmdIdxTrie) { t.value = v },
+		func(t *SubCmdIdxTrie, indices SubCmdIdx) {
+			if t.nextLevel == nil {
+				t.nextLevel = map[uint64]*SubCmdIdxTrie{}
+			}
+			if _, ok := t.nextLevel[indices[0]]; !ok {
+				t.nextLevel[indices[0]] = &SubCmdIdxTrie{}
+			}
+		}, nil)
+}
+
+// RemoveValue tries to remove a value indexed by the given SubCmdIdx in the
+// trie. If a value is found, removes it and returns true. If a value with that
+// SubCmdIdx is not found, returns false.
+func (t *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
+	hadValue := false
+	r := t.index(indices, func(t *SubCmdIdxTrie) {
```

Comment:
Due to the closures, captured value, order of calls, I find this exceptionally hard to read and follow.
Here's an alternative implementation that passes the tests:

```go

// SubCmdIdxTrie is a map-based trie using SubCmdIdx for indexing the data
// stored inside.
type SubCmdIdxTrie struct {
	value    interface{}
	children map[uint64]*SubCmdIdxTrie
}

// Value returnes the value stored in the trie indexed by the given SubCmdIdx.
// if no value is found by the given SubCmdIdx, returns nil.
func (t *SubCmdIdxTrie) Value(indices SubCmdIdx) interface{} {
	for t != nil && len(indices) > 0 && len(t.children) > 0 {
		t, indices = t.children[indices[0]], indices[1:]
	}
	if t == nil || len(indices) > 0 {
		return nil // invalid index
	}
	return t.value
}

// SetValue sets a value to the trie with the given SubCmdIdx as index.
func (t *SubCmdIdxTrie) SetValue(indices SubCmdIdx, v interface{}) {
	if v == nil { // nil value behaves the same as no value
		t.RemoveValue(indices)
		return
	}
	for _, i := range indices {
		next, ok := t.children[i]
		if !ok { // child does not exist - create it
			if t.children == nil {
				t.children = map[uint64]*SubCmdIdxTrie{}
			}
			next = &SubCmdIdxTrie{}
			t.children[i] = next
		}
		t = next
	}
	t.value = v
}

// RemoveValue tries to remove a value indexed by the given SubCmdIdx in the
// trie. If a value is found, removes it and returns true. If a value with that
// SubCmdIdx is not found, returns false.
func (t *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
	n := len(indices)
	stack := make([]*SubCmdIdxTrie, n+1)
	stack[0] = t
	for i, j := range indices { // build a stack of nodes starting with t
		t = t.children[j]
		if t == nil {
			return false // invalid index
		}
		stack[i+1] = t
	}
	if t.value == nil {
		return false // cannot remove a value it doesn't have
	}
	t.value = nil             // clear the value
	for i := n; i >= 0; i-- { // remove nodes that have no value or children
		if t.value != nil || len(t.children) > 0 {
			break // node cannot be deleted
		}
		delete(stack[i].children, indices[i-1])
		t = stack[i]
	}
	return true
}
```

This implementation is fractionally faster too:

Old:
```bash
$ go test ./gapis/api/ --bench=.                                                                                                                          
BenchmarkSubCmdIdxTrie-48         100000             11770 ns/op 
```

New:
```bash
$ go test ./gapis/api/ --bench=.                                                                                                                                                    
BenchmarkSubCmdIdxTrie-48         200000             10194 ns/op
```